exporters/metric/dogstatsd/internal/statsd/conn.go

```diff
@@ -0,0 +1,291 @@
+		conn    net.Conn
+		writer  io.Writer
+		buffer  bytes.Buffer
+	}
+
+	// Adapter supports statsd syntax variations, primarily plain
+	// statsd vs. dogstatsd.
+	Adapter interface {
+		AppendName(export.Record, *bytes.Buffer)
+		AppendTags(export.Record, *bytes.Buffer)
+	}
+)
+
+const (
+	formatCounter   = "c"
+	formatHistogram = "h"
+	formatGauge     = "g"
+	formatTiming    = "ms"
+
+	// MaxPacketSize defaults to Ethernet Jumbo frame size.  If
+	// this is too large, you will see "write: message too long"
+	// errors.
+	MaxPacketSize = 9216
+)
+
+var (
+	_ export.Exporter = &Exporter{}
+
+	ErrInvalidScheme = fmt.Errorf("invalid statsd transport")
+)
+
+// NewExporter returns a common implementation for exporters that Export
+// statsd syntax.
+func NewExporter(config Config, adapter Adapter) (*Exporter, error) {
+	if config.MaxPacketSize <= 0 {
+		config.MaxPacketSize = MaxPacketSize
+	}
+	var writer io.Writer
+	var conn net.Conn
+	var err error
+	if config.Writer != nil {
+		writer = config.Writer
+	} else {
+		conn, err = dial(config.URL)
+		if conn != nil {
+			writer = conn
+		}
+	}
+	// TODO: If err != nil, we return it _with_ a valid exporter; the
+	// exporter should attempt to re-dial if it's retryable.  Add a
+	// Start() and Stop() API.
+	return &Exporter{
+		adapter: adapter,
+		config:  config,
+		conn:    conn,
+		writer:  writer,
+	}, err
+}
+
+// dial connects to a statsd service using several common network
+// types.  Presently "udp" and "unix" datagram socket connections are
+// supported.
+func dial(endpoint string) (net.Conn, error) {
+	dest, err := url.Parse(endpoint)
+	if err != nil {
+		return nil, err
+	}
+
+	// TODO: Support tcp destination, need configurable timeouts first.
+
+	scheme := dest.Scheme
+	switch scheme {
+	case "udp", "udp4", "udp6":
+		udpAddr, err := net.ResolveUDPAddr(scheme, dest.Host)
+		locAddr := &net.UDPAddr{}
+		if err != nil {
+			return nil, err
+		}
+		conn, err := net.DialUDP(scheme, locAddr, udpAddr)
+		if err != nil {
+			return nil, err
+		}
+		return conn, err
+	case "unix", "unixgram":
+		scheme = "unixgram"
+		locAddr := &net.UnixAddr{}
+
+		sockAddr, err := net.ResolveUnixAddr(scheme, dest.Path)
+		if err != nil {
+			return nil, err
+		}
+		conn, err := net.DialUnix(scheme, locAddr, sockAddr)
+		if err != nil {
+			return nil, err
+		}
+		return conn, err
+	}
+	return nil, ErrInvalidScheme
+}
+
+// Export is common code for any statsd-based metric.Exporter implementation.
+func (e *Exporter) Export(_ context.Context, checkpointSet export.CheckpointSet) error {
+	buf := &e.buffer
+	buf.Reset()
+
+	var aggErr error
+	var sendErr error
+
+	aggErr = checkpointSet.ForEach(func(rec export.Record) error {
+		before := buf.Len()
+
+		if err := e.formatMetric(rec, buf); err != nil {
+			return err
+		}
+
+		if buf.Len() < e.config.MaxPacketSize {
+			return nil
+		}
+		if before == 0 {
+			// A single metric >= packet size
+			if err := e.send(buf.Bytes()); err != nil && sendErr == nil {
```

Comment:
I'm still running into the same issue with this line of code....

my console is full of ```Metrics SDK error: write udp 127.0.0.1:62433->127.0.0.1:8125: write: message too long```

a breakpoint on this line shows buf has a len of 10791, and is essentially:

```
example.histogram:0|h|#climb:every_200ms,kube_pod:Chriss-MacBook-Pro.local,service:o11y-sensor-calibration
example.histogram:1|h|#climb:every_200ms,kube_pod:Chriss-MacBook-Pro.local,service:o11y-sensor-calibration
example.histogram:3|h|#climb:every_200ms,kube_pod:Chriss-MacBook-Pro.local,service:o11y-sensor-calibration
...
example.histogram:97|h|#climb:every_200ms,kube_pod:Chriss-MacBook-Pro.local,service:o11y-sensor-calibration
example.histogram:98|h|#climb:every_200ms,kube_pod:Chriss-MacBook-Pro.local,service:o11y-sensor-calibration
```

I tried tracing this code before and I think an assumption is off.